ymer/src/main/java/com/avanza/ymer/BulkMirroredObjectWriter.java

```diff
@@ -0,0 +1,167 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.avanza.ymer;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.bson.Document;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.gigaspaces.sync.DataSyncOperation;
+import com.gigaspaces.sync.OperationsBatchData;
+import com.mongodb.MongoBulkWriteException;
+import com.mongodb.bulk.BulkWriteError;
+import com.mongodb.bulk.BulkWriteResult;
+
+final class BulkMirroredObjectWriter {
+
+	private static final Logger logger = LoggerFactory.getLogger(BulkMirroredObjectWriter.class);
+
+	private final SpaceMirrorContext mirror;
+	private final DocumentWriteExceptionHandler exceptionHandler;
+	private final MirroredObjectFilterer objectFilterer;
+
+	BulkMirroredObjectWriter(SpaceMirrorContext mirror,
+			DocumentWriteExceptionHandler exceptionHandler,
+			MirroredObjectFilterer objectFilterer
+	) {
+		this.mirror = requireNonNull(mirror);
+		this.exceptionHandler = requireNonNull(exceptionHandler);
+		this.objectFilterer = requireNonNull(objectFilterer);
+	}
+
+	public void executeBulk(InstanceMetadata metadata, OperationsBatchData batch) {
+		Map<String, List<MongoBulkChange>> changesByCollection = new HashMap<>();
+
+		for (DataSyncOperation bulkItem : objectFilterer.filterSpaceObjects(batch.getBatchDataItems())) {
+			String collectionName = mirror.getCollectionName(bulkItem.getDataAsObject().getClass());
+			List<MongoBulkChange> bulkChanges = changesByCollection.computeIfAbsent(collectionName, x -> new ArrayList<>());
+
+			switch (bulkItem.getDataSyncOperationType()) {
+				case WRITE:
+					bulkChanges.add(new MongoBulkChange(MirrorOperation.INSERT, bulkItem.getDataAsObject()));
+					break;
+				case UPDATE:
+				case PARTIAL_UPDATE:
+					bulkChanges.add(new MongoBulkChange(MirrorOperation.UPDATE, bulkItem.getDataAsObject()));
+					break;
+				case REMOVE:
+					bulkChanges.add(new MongoBulkChange(MirrorOperation.REMOVE, bulkItem.getDataAsObject()));
+					break;
+				default:
+					throw new UnsupportedOperationException("Bulkoperation " + bulkItem.getDataSyncOperationType() + " is not supported");
+			}
+		}
+
+		changesByCollection.forEach((key, value) -> new MongoBulkWriter(metadata, key).execute(value));
```

Comment:
I was thinking we could do this in parallel, what do you guys think?

We should be able to perform updates for different collections in parallel without any issues, right?

Reply:
I think it would be possible, yes. I'm not sure however how big the performance gain would be. I think we often have one of two cases: either a PU with lots of writes to one collection or few writes to several collections ...

Reply:
Yeah, maybe it just adds unnecessary complexity.